plugin/hooks.go

```diff
@@ -149,4 +154,41 @@ type Hooks interface {
 	// Note that this method will be called for files uploaded by plugins, including the plugin that uploaded the post.
 	// FileInfo.Size will be automatically set properly if you modify the file.
 	FileWillBeUploaded(c *Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string)
+
+	// PushNotificationWillBeSent is invoked before the push notification has been sent to the push proxy server.
+	// You should return the notification that you want to be sent to the user (normally the same notification).
+	//
+	// You can reject the notification returning a nil notification.
+	//
+	// You can modify the notification, returning a changed version of the notification.
+	//
+	// Minimum server version: 5.10
+	PushNotificationWillBeSent(c *Context, notification *model.PushNotification) *model.PushNotification
+
+	// PushNotificationHasBeenSent is invoked after the push notification has been sent to the push proxy server.
+	//
+	// Minimum server version: 5.10
+	PushNotificationHasBeenSent(c *Context, notification *model.PushNotification)
+
+	// PushNotificationHasFailed is invoked after the push proxy server has replied with an error.
+	//
+	// Minimum server version: 5.10
+	PushNotificationHasFailed(c *Context, notification *model.PushNotification, err error)
+
+	// PushNotificationEnqueued is invoked just before the notification has been enqueued to be sent to the push proxy server.
+	//
+	// This is an internal hook and it can change anytime in the future.
+	//
+	// Minimum server version: 5.10
+	PushNotificationEnqueued(c *Context, notificationId, notificationType, userId, channelId, postId string)
+
+	// PushNotificationAck is invoked when the server receive the confirmation
+	// of the reception of the push notification from the mobile device.
+	//
+	// You can return a notification that will be return to the mobile device
+	// as response to the ACK, and the device will update the notification
+	// stored in the device.
+	//
+	// Minimum server version: 5.10
+	PushNotificationAck(c *Context, ackId string, recievedAt, ackAt int64) *model.PushNotification
```

Comment:
I would opt to mark this also as internal. 1/5

Reply:
I don't see that, this open a lot of possibilities for plugin developers, like modify the notification on ack. For example for enrich notifications.

Reply:
I don't have a strong opinion on this. Fine with the current way.